x/valset-pref/validator_set_test.go

```diff
@@ -274,3 +205,132 @@ func (s *KeeperTestSuite) TestIsPreferenceValid() {
 		})
 	}
 }
+
+// NOTE: this is the case that used to error. Fixed by this PR
+func (s *KeeperTestSuite) TestUndelegateFromValSetErrorCase() {
+	s.SetupTest()
+
+	valAddrs := s.SetupMultipleValidators(2)
+	valPreferences := []types.ValidatorPreference{
+		{
+			ValOperAddress: valAddrs[0],
+			Weight:         sdk.NewDecWithPrec(5, 1), // 0.5
+		},
+		{
+			ValOperAddress: valAddrs[1],
+			Weight:         sdk.NewDecWithPrec(5, 1), // 0.5
+		},
+	}
+
+	delegator := sdk.AccAddress([]byte("addr1---------------"))
+	coinToStake := sdk.NewCoin(sdk.DefaultBondDenom, sdk.NewInt(10_000_000))   // delegate 10osmo using Valset now and 10 osmo using regular staking delegate
+	coinToUnStake := sdk.NewCoin(sdk.DefaultBondDenom, sdk.NewInt(20_000_000)) // undelegate 20osmo
+	expectedShares := []sdk.Dec{sdk.NewDec(15_000_000), sdk.NewDec(500_000)}
+
+	s.FundAcc(delegator, sdk.Coins{sdk.NewInt64Coin(sdk.DefaultBondDenom, 100_000_000)}) // 100 osmo
+
+	// valset test setup
+	// SetValidatorSetPreference sets a new list of val-set
+	_, err := s.App.ValidatorSetPreferenceKeeper.DeriveValidatorSetPreference(s.Ctx, delegator.String(), valPreferences)
+	s.Require().NoError(err)
+
+	s.App.ValidatorSetPreferenceKeeper.SetValidatorSetPreferences(s.Ctx, delegator.String(), types.ValidatorSetPreferences{
+		Preferences: valPreferences,
+	})
+
+	// DelegateToValidatorSet delegate to existing val-set
+	err = s.App.ValidatorSetPreferenceKeeper.DelegateToValidatorSet(s.Ctx, delegator.String(), coinToStake)
+	s.Require().NoError(err)
+
+	valAddr, err := sdk.ValAddressFromBech32(valAddrs[0])
+	s.Require().NoError(err)
+
+	validator, found := s.App.StakingKeeper.GetValidator(s.Ctx, valAddr)
+	s.Require().True(found)
+
+	// Delegate more token to the validator. This will cause valset and regular staking to go out of sync
+	_, err = s.App.StakingKeeper.Delegate(s.Ctx, delegator, sdk.NewInt(10_000_000), stakingtypes.Unbonded, validator, true)
+	s.Require().NoError(err)
+
+	err = s.App.ValidatorSetPreferenceKeeper.UndelegateFromValidatorSet(s.Ctx, delegator.String(), coinToUnStake)
+	s.Require().NoError(err)
+
+	for i, val := range valPreferences {
+		valAddr, err := sdk.ValAddressFromBech32(val.ValOperAddress)
+		s.Require().NoError(err)
+
+		// guarantees that the delegator exists because we check it in UnDelegateToValidatorSet
+		del, found := s.App.StakingKeeper.GetDelegation(s.Ctx, delegator, valAddr)
+		if found {
+			s.Require().Equal(expectedShares[i], del.GetShares())
+		}
+	}
```

Comment:
I might be reading this wrong, but it seems like we are just checking delegation and not undelegating funds right? Also, shouldn't we require found to be true here? I worry that we could skip the require equal and not know it if the delegation was not found.

Reply:
on second thought, i realized that it's better to check for `Undelegating` tokens, so refactored the validation check here: https://github.com/osmosis-labs/osmosis/pull/5967/files#diff-e8d94c443b26e1cda354d150b30ce5542344099e0bbbb8e84e507d9df791daf7R513-R532